packages/controllers/src/snaps/SnapController.ts

```diff
@@ -698,6 +773,133 @@ export class SnapController extends BaseController<
+    }
+
+    const blockedSnaps = await this._checkSnapBlockList(
+      Object.values(this.state.snaps).reduce<Record<SnapId, SemVerVersion>>(
+        (blockListArg, snap) => {
+          blockListArg[snap.id] = snap.version;
+          return blockListArg;
+        },
+        {},
+      ),
+    );
+
+    await Promise.all(
+      Object.entries(blockedSnaps).map(
+        ([snapId, { blocked, infoUrl, reason }]) => {
+          if (blocked) {
+            return this._blockSnap(snapId, { infoUrl, reason });
+          }
+
+          this._unblockSnap(snapId);
+          return Promise.resolve();
+        },
+      ),
+    );
+  }
+
+  /**
+   * Blocks an installed snap and prevents it from being started again. Emits
+   * {@link SnapBlocked}. Does nothing if the snap is not installed.
+   *
+   * @param snapId - The snap to block.
+   * @param blockedSnapInfo - Information detailing why the snap is blocked.
+   */
+  private async _blockSnap(
+    snapId: SnapId,
+    blockedSnapInfo: BlockedSnapInfo,
+  ): Promise<void> {
+    if (!this.has(snapId)) {
+      return;
+    }
+
+    try {
+      this.update((state: any) => {
+        state.snaps[snapId].blocked = true;
+        state.snaps[snapId].blockInformation = blockedSnapInfo;
+      });
+
+      await this.disableSnap(snapId);
+    } catch (error) {
+      console.error(
+        `Encountered error when stopping blocked snap "${snapId}".`,
+        error,
+      );
+    }
+
+    this.messagingSystem.publish(
+      `${controllerName}:snapBlocked`,
+      snapId,
+      blockedSnapInfo,
+    );
+  }
+
+  /**
+   * Unblocks a snap so that it can be enabled and started again. Emits
+   * {@link SnapUnblocked}. Does nothing if the snap is not installed or already
+   * unblocked.
+   *
+   * @param snapId - The id of the snap to unblock.
+   */
+  private async _unblockSnap(snapId: SnapId): Promise<void> {
+    if (!this.has(snapId) || !this.state.snaps[snapId].blocked) {
+      return;
+    }
+
+    this.update((state: any) => {
+      state.snaps[snapId].blocked = false;
+      delete state.snaps[snapId].blockInformation;
+    });
+
+    this.messagingSystem.publish(`${controllerName}:snapUnblocked`, snapId);
+  }
+
+  /**
+   * Checks the block list to determine whether a version of a snap is blocked.
+   *
+   * @param snapId - The snap id to check.
+   * @param version - The version of the snap to check.
+   * @returns Whether the version of the snap is blocked or not.
+   */
+  async isBlocked(
+    snapId: ValidatedSnapId,
+    version: SemVerVersion,
+  ): Promise<boolean> {
+    if (!this._checkSnapBlockList) {
+      throw new Error(BLOCK_LIST_UNDEFINED_ERROR);
+    }
+
+    const result = await this._checkSnapBlockList({ [snapId]: version });
+    return result[snapId].blocked;
+  }
+
+  /**
+   * Asserts that a version of a snap is not blocked. Succeeds automatically
+   * if {@link SnapController._checkSnapBlockList} is undefined.
```

Comment:
How come we want it to succeed automatically when `_checkSnapBlockList` is undefined when we throw elsewhere if it is undefined? Do we require it to be defined or not :sweat_smile:

Reply:
I agree that `assertIsUnblocked` should have same logic as `isBlocked`. One throws, other succeeds.

`isBlocked()` could return `Promise<boolean | null>` it we want to pass that information up

Reply:
I just removed the `checkBlockList` check so that it now mirrors `isBlocked`.